### PR TITLE
Implemented requested functionality

### PR DESCRIPTION
**Pull request by PavloSliepchenko**

### CODE REVIEW

File: src/main/java/mate/academy/rickandmorty/Application.java [hunk 1]
```diff
@@ -1,12 +1,24 @@
...
+    private DataDownLoadService downLoadService;
 
     public static void main(String[] args) {
         SpringApplication.run(Application.class, args);
+        downLoadServiceStatic.getAllCharactersFromApi();
+    }
+
+    @PostConstruct
+    private void setDownLoadServiceStatic() {
+        downLoadServiceStatic = this.downLoadService;
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/Application.java]: Dont complicate main class with extra logic. 
See CommandLineRunner or ApplicationRunner for loading data on startup.
```suggestion
```

File: src/main/java/mate/academy/rickandmorty/controller/Controller.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping(value = "/api/characters")
+@RequiredArgsConstructor
+@Tag(name = "Api for searching characters of Rick and Morty series",
+        description = "Helps to find characters by name or provides a random character")
+public class Controller {
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/controller/Controller.java]: Chose better name for controller. 
Just `Controller` is too generic. 
```suggestion
public class Controller {
```

File: src/main/java/mate/academy/rickandmorty/mapper/Mapper.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package mate.academy.rickandmorty.mapper;
+
+import mate.academy.rickandmorty.dto.CharacterResponseDto;
+import mate.academy.rickandmorty.model.Character;
+import org.springframework.stereotype.Component;
+
+@Component
+public class Mapper {
+    public Character responseDtoToCharacter(CharacterResponseDto responseDto) {
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/mapper/Mapper.java]: just to you: consider using Mapstruct

File: src/main/java/mate/academy/rickandmorty/repository/Repository.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package mate.academy.rickandmorty.repository;
+
+import java.util.List;
+import mate.academy.rickandmorty.model.Character;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface Repository extends JpaRepository<Character, Long> {
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/repository/Repository.java]: Repository is bad name. Too generic. 
```suggestion
public interface Repository extends JpaRepository<Character, Long> {
```

File: src/main/java/mate/academy/rickandmorty/service/ClientService.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import mate.academy.rickandmorty.repository.Repository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ClientService {
+    private static final long NUMBER_OF_CHARACTERS = 20;
+    @Autowired
+    private Repository repository;
+    private Random random = new Random();
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/service/ClientService.java]: We are using the spring, so let's do everything in spring style. 
make the bean of random. 
```suggestion
   @Autowired private Random random;
```

File: src/main/java/mate/academy/rickandmorty/service/ClientService.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+    public Character getRandomCharacter() {
+        Long randomCharacter = random.nextLong(NUMBER_OF_CHARACTERS) + 1;
+        return repository.findById(randomCharacter).get();
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/service/ClientService.java]: Direct use of `Optional#get` without `Optional#isPresent` is bad practice. 

Consider using the Optional#orElse, OPtional#orELseThrow, or null check before Optional#get.
```suggestion
        return repository.findById(randomCharacter).get();
```

File: src/main/java/mate/academy/rickandmorty/service/DataDownLoadService.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+@RequiredArgsConstructor
+public class DataDownLoadService {
+    private static final String URL = "https://rickandmortyapi.com/api/character/";
+    private static final int NUMBER_OF_CHARACTERS = 20;
+    private final Mapper mapper;
+    private final ObjectMapper jasonMapper;
+    private final Repository repository;
+
+    public void getAllCharactersFromApi() {
+        HttpClient client = HttpClient.newHttpClient();
```
Comment (sokimaaa) [inline src/main/java/mate/academy/rickandmorty/service/DataDownLoadService.java]: 1 You instantiate the client after each method invocation... Fix it. 

2. Consider using the spring style (create a bean and inject it throughout your app)
```suggestion
        HttpClient client = HttpClient.newHttpClient();
```

File: src/main/resources/application.properties [hunk 8]
```diff
@@ -1 +1,6 @@
-
+spring.datasource.url=jdbc:mysql://localhost:3306/rick_and_morty
+spring.datasource.username=root
+spring.datasource.password=Passw0rdRO0Tpassvvord
+spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
```
Comment (sokimaaa) [inline src/main/resources/application.properties:4]: Is this property required? 
if not should be removed. 
It already deprecated and may cause a bunch of errors.
```suggestion
```

**Review by sokimaaa — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/rickandmorty/init/InitDataBase.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+import mate.academy.rickandmorty.exception.DataProcessException;
+import mate.academy.rickandmorty.mapper.DtoMapper;
+import mate.academy.rickandmorty.model.Character;
+import mate.academy.rickandmorty.repository.CharacterRepository;
+import org.springframework.boot.CommandLineRunner;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class InitDataBase implements CommandLineRunner {
```
Comment (koval-kn) [inline src/main/java/mate/academy/rickandmorty/init/InitDataBase.java]: your name sounds more like method name rather class name
by the way, where's your Application class?
```suggestion
public class DataInitializer implements CommandLineRunner {
```

File: src/main/java/mate/academy/rickandmorty/controller/CharactersController.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping(value = "/api/characters")
+@RequiredArgsConstructor
+@Tag(name = "Api for searching characters of Rick and Morty series",
+        description = "Helps to find characters by name or provides a random character")
+public class CharactersController {
```
Comment (koval-kn) [inline src/main/java/mate/academy/rickandmorty/controller/CharactersController.java]: lets use singular form 
```suggestion
public class CharacterController {
```

File: src/main/java/mate/academy/rickandmorty/service/ClientService.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import mate.academy.rickandmorty.repository.CharacterRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ClientService {
+    private static final long NUMBER_OF_CHARACTERS = 20;
+    @Autowired
+    private CharacterRepository characterRepository;
+    private final Random random = new Random();
```
Comment (koval-kn) [inline src/main/java/mate/academy/rickandmorty/service/ClientService.java]: not fixed prev comment about using spring to inject instance

File: src/main/java/mate/academy/rickandmorty/service/ClientService.java [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+import mate.academy.rickandmorty.model.Character;
+import mate.academy.rickandmorty.repository.CharacterRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ClientService {
+    private static final long NUMBER_OF_CHARACTERS = 20;
+    @Autowired
+    private CharacterRepository characterRepository;
```
Comment (koval-kn) [inline src/main/java/mate/academy/rickandmorty/service/ClientService.java]: use  `@RequiredArgsConstructor` for constructor-based injection, it's better way
```suggestion
    private final CharacterRepository characterRepository;
```

File: src/main/java/mate/academy/rickandmorty/init/InitDataBase.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    private static final String URL = "https://rickandmortyapi.com/api/character/";
+    private static final int NUMBER_OF_CHARACTERS = 20;
+    private final DtoMapper mapper;
+    private final ObjectMapper jasonMapper;
+    private final CharacterRepository characterRepository;
+
```
Comment (koval-kn) [inline src/main/java/mate/academy/rickandmorty/init/InitDataBase.java]: not fixed previous comment 
need to create `HttpClient` bean in your Spring Configuration
```suggestion
private final HttpClient httpClient;

```

**Review by koval-kn — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/rickandmorty/init/InitDataBase.java (hunk 13 above)
Comment (PavloSliepchenko) [reply src/main/java/mate/academy/rickandmorty/init/InitDataBase.java]: This method runs only once when the application started, isn't it?

File: src/main/java/mate/academy/rickandmorty/init/InitDataBase.java (hunk 9 above)
Comment (PavloSliepchenko) [reply src/main/java/mate/academy/rickandmorty/init/InitDataBase.java]: Here it is
Screenshot 2023-11-07 at 8 02 05 AM

File: src/main/java/mate/academy/rickandmorty/init/DataBaseInitializer.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    private static final int NUMBER_OF_CHARACTERS = 20;
+    private final DtoMapper mapper;
+    private final ObjectMapper jasonMapper;
+    private final CharacterRepository characterRepository;
+    private final HttpClient client;
+
+    @Override
+    public void run(String... args) {
+        List<Character> characters = new ArrayList<>();
+        for (int i = 1; i <= NUMBER_OF_CHARACTERS; i++) {
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/init/DataBaseInitializer.java]: Where do you get NUMBER_OF_CHARACTERS from? Why 20?

image

File: src/main/java/mate/academy/rickandmorty/controller/CharacterController.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    @GetMapping(value = "/random")
+    @Operation(summary = "Random character", description = "Provides a random character")
+    public Character randomCharacter() {
+        return clientService.getRandomCharacter();
+    }
+
+    @GetMapping
+    @Operation(summary = "Get a character by name",
+            description = "Provides a list of characters by providing string")
+    public List<Character> getCharactersBySequence(String name) {
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/controller/CharacterController.java]: RequestParam?

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Review by koval-kn — APPROVED** [review]

**Pull request closed without merging**